### PR TITLE
cmake/toolchain: add support gnu arc embedded toolchain

### PR DESCRIPTION
**Pull request by vonhust**
Although arc gnu toolchain is intergrated into Zephyr SDK,
here we add the support of arc gnu toolchain release separately
by synopsys in
https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain
because:

* Zephyr SDK only supports Linux, Synopsys ARC GNU toolchain supports
  both Windows and Linux. So ARC users can use Zephyr in windows.

* Useful for development and test of new features, bug fixes of ARC GNU
  toolchain before intergrated into Zephyr SDK in up stream.

To use synopsys arc gnu toolchain, you need:
   * set ZEPHYR_TOOLCHAIN_VARIANT to gnuarc
   * set GNUARC_TOOLCHAIN_PATH to the folder where synopsys arc gnu
     toolchain is located. For example if arc-elf32-gcc is in
     /opt/arc_gnu/bin/, then set GNUARC_TOOLCHAIN_PATH to
     /opt/arc_gnu

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

### CODE REVIEW

**Comment by carlescufi** [comment]
>set ZEPHYR_TOOLCHAIN_VARIANT to gnuarcemb

Should we call this `gnuarc` instead? There's no reference to "'embedded" in https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain

>set GNUARCEMB_TOOLCHAIN_PATH

Similarly, perhaps `GNUARC_TOOLCHAIN_PATH`

**Review by stephanosio — COMMENTED** [review]
Looks ok overall.

As @carlescufi suggested, I think it would be better/more intuitive to call this `gnuarc`. `gnuarmemb` was named as such because the toolchain distribution from ARM is called "GNU ARM Embedded."

p.s. also `gnuarmemb` and `gnuarcemb` look too similar, so this can easily lead to confusion/misreading/typo.

**Comment by vonhust** [comment]
> > set ZEPHYR_TOOLCHAIN_VARIANT to gnuarcemb
> 
> Should we call this `gnuarc` instead? There's no reference to "'embedded" in https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain
> 
> > set GNUARCEMB_TOOLCHAIN_PATH
> 
> Similarly, perhaps `GNUARC_TOOLCHAIN_PATH`

thanks  for advice,  i will change it to GNUARC

File: cmake/toolchain/gnuarc/Kconfig [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# GNU ARC Toolchain
+
+# Copyright (c) 2020 Synopsys
+# SPDX-License-Identifier: Apache-2.0
+
+config TOOLCHAIN_GNUARCEMB
```
Comment (stephanosio) [inline cmake/toolchain/gnuarc/Kconfig:6]: missed this one:

```suggestion
config TOOLCHAIN_GNUARC
```

Comment (vonhust) [reply cmake/toolchain/gnuarc/Kconfig:6]: Good catch. Thanks

**Comment by abrodkin** [comment]
My only concern here is why do we need all that per-architecture nonsense? IMHO as along as we're talking about GCC-based toolchains (which is true for ARC & ARM ones, right?) we only need to have its `bin` folder in the `PATH` and `CROSS_COMPILE` set accordingly.

I guess there's a good reason why this was not done before like that:
1. Unset `ZEPHYR_TOOLCHAIN_VARIANT`
2. Set `CROSS_COMPILE` to your liking

That simple.

**Comment by stephanosio** [comment]
> My only concern here is why do we need all that per-architecture nonsense? IMHO as along as we're talking about GCC-based toolchains (which is true for ARC & ARM ones, right?) we only need to have its `bin` folder in the `PATH` and `CROSS_COMPILE` set accordingly.

I think you are missing two things here:

* Compiler toolchain-specific properties (e.g. `nano.specs` support is not universal across all GCC-based toolchain distributions)
* Tools other than the GNU binutils and gcc (e.g. openocd)

This toolchain spec resolves and uses all the right tools once you select it and this results in a better developer experience overall.

**Comment by abrodkin** [comment]
> 
> 
> > My only concern here is why do we need all that per-architecture nonsense? IMHO as along as we're talking about GCC-based toolchains (which is true for ARC & ARM ones, right?) we only need to have its `bin` folder in the `PATH` and `CROSS_COMPILE` set accordingly.
> 
> I think you are missing two things here:
> 
>     * Compiler toolchain-specific properties (e.g. `nano.specs` support is not universal across all GCC-based toolchain distributions)

Ok that's true, but I didn't see we specify it in this PR, do we?

>     * Tools other than the GNU binutils and gcc (e.g. openocd)

Not sure about other pre-built GNU toolchains, but the one for ARC doesn't have OpenOCD binary bundled into it, instead it's only a part of our "GNU IDE", so as of today it's not available in pre-built form separately. Moreover given Zephyr's OpenOCD has a couple of important fixes & improvements added on top of upstream version I'd strongly suggest people to use OpenOCD bundled in Zephyr's host tools (which are now available separately from the rest of toolchains).

So in that sense setup of `OPENOCD_DEFAULT_PATH` for `gnuarc` toolchain shouldn't be done and there should be a way to still use Zephyr's host tool which BTW also include now QEMU for ARC and that [I mean QEMU] probably won't be distributed by us in pre-built form because it's not clear if it adds any value.

**Comment by stephanosio** [comment]
> Ok that's true, but I didn't see we specify it in this PR, do we?

https://github.com/zephyrproject-rtos/zephyr/pull/25931/files#diff-01d722867c1ef63cc087e7ed116237a2R8
https://github.com/zephyrproject-rtos/zephyr/pull/25931/files#diff-11b1a5e8f13b46a629478e560ec0f37aR22

> Not sure about other pre-built GNU toolchains, but the one for ARC doesn't have OpenOCD binary bundled into it, instead it's only a part of our "GNU IDE", so as of today it's not available in pre-built form separately.

As far as I can see, this "toolchain spec" is actually for the IDE.

**Comment by nashif** [comment]
I tend to agree with @abrodkin, there might be some special casing required for some of the toolchains, but when dealing with a gcc based toolchains we probably want to have this covered with one variant only, gnugcc or whatever and cover any toolchain that is based on gcc using one variant and only requiring the path to to ../bin/<foo-bar>-
This will give us more coverage and support for gcc in one place. I do not think tooling (openocd, etc) is part of this, this is orthogonal

**Comment by stephanosio** [comment]
> Moreover given Zephyr's OpenOCD has a couple of important fixes & improvements added on top of upstream version I'd strongly suggest people to use OpenOCD bundled in Zephyr's host tools (which are now available separately from the rest of toolchains).
> 
> So in that sense setup of `OPENOCD_DEFAULT_PATH` for `gnuarc` toolchain shouldn't be done and there should be a way to still use Zephyr's host tool which BTW also include now QEMU for ARC and that [I mean QEMU] probably won't be distributed by us in pre-built form because it's not clear if it adds any value.

While I agree with that for the most part, I am not sure if that will be feasible for everyone (e.g. what if your company requires you to use an "official" toolchain distribution from Synopsys?).

But, I am sure you guys have better understanding of this, so I won't comment on it any further.

**Comment by stephanosio** [comment]
> This will give us more coverage and support for gcc in one place. I do not think tooling (openocd, etc) is part of this, this is orthogonal

@nashif As far as I can see, the definition of "toolchain" includes the host tools (e.g. Zephyr SDK).

**Comment by abrodkin** [comment]
@stephanosio if indeed [and that seems to be the case] Zephyr's "toolchain" includes those 3 distinct components (which I'd say are not typically a part of GNU/GCC toolchain - I might be wrong here as I didn't do any research among all the arches) we'd need either to bundle all of them in our `snps-gnu` `VARIANT` or somehow still use some parts of default Zephyr's "toolchain". But I'm not sure if we are going to bundle more things (like even QEMU in the our GNU IDE, though it might turn out to be not the worst idea ever) in our one and only prebuilt toolchain just for Zephyr. And even if we do what about other vendors?

So I'd prefer more UNIX-ish way with ability to select components of choice... or rather override those I really need/have to.

As for newlib version it's a good question but frankly as usual with OS-es I'd prefer to use OS' built-in libc implementation so that I get predictable result instead of long outdated newlib from the external toolchain. This BTW is also applicable to libgcc primitives - it's good to have them inside the OS but it comes with a price I know :)

**Comment by stephanosio** [comment]
> @stephanosio if indeed [and that seems to be the case] Zephyr's "toolchain" includes those 3 distinct components (which I'd say are not typically a part of GNU/GCC toolchain - I might be wrong here as I didn't do any research among all the arches) we'd need either to bundle all of them in our `snps-gnu` `VARIANT` or somehow still use some parts of default Zephyr's "toolchain". But I'm not sure if we are going to bundle more things (like even QEMU in the our GNU IDE, though it might turn out to be not the worst idea ever) in our one and only prebuilt toolchain just for Zephyr. And even if we do what about other vendors?

While I do not want to confuse people with what is not yet ready for public testing, I think it is worth mentioning here -- I believe this approach should aid in making "compiler toolchain" and "host tools" specification/selection orthogonal: https://github.com/stephanosio/zephyr-crosstool-ng/issues/25.

With this approach, for example, people can use the GCC-based compiler toolchains through `ZEPHYR_TOOLCHAIN_VARIANT=cross-compile` while using the Zephyr host tools such as OpenOCD and QEMU, as they are automatically aliased/shimmed into the `PATH` location when the packages are installed; this is of course controllable by the package manager.

**Comment by abrodkin** [comment]
@stephanosio that's awesome! And especially in that light do we want to introduce yet another toolchain variant now and probably remove it later as being pretty much incompatible with newer Zephyr?

**Comment by vonhust** [comment]
Close this PR, as we can use CROSS_COMPILE to enable arc gnu.  when it's necessary to have a specific
ARC GNU support,  submit a PR again.  Meanwhile,  some docs of add ARC GNU through CROSS_COMPILE will be added in another PR.

**Pull request closed without merging**